src/core/thread/wakeup_tx_scheduler.cpp

```diff
@@ -0,0 +1,192 @@
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "wakeup_tx_scheduler.hpp"
+
+#if OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/log.hpp"
+#include "common/num_utils.hpp"
+#include "common/time.hpp"
+#include "core/instance/instance.hpp"
+#include "radio/radio.hpp"
+
+#include <openthread/platform/radio.h>
+
+namespace ot {
+
+namespace {
+
+// Frame lengths including SHR
+constexpr uint32_t kWakeupFrameLength   = 54;
+constexpr uint32_t kParentRequestLength = 78;
+
+// A rough estimate of the size of the data that has to be exchanged over the Radio Co-Processor transport to schedule
+// a wake-up frame transmission by the RCP. This is used to make sure that a wake-up frame is received by the RCP early
+// enough to be transmitted on time.
+constexpr uint32_t kWakeupFrameWeight = 100;
+
+uint16_t CalcTxRequestAheadTimeUs(Instance &aInstance)
+{
+    constexpr uint32_t kBitsPerByte = 8;
+    constexpr uint32_t kUsPerSecond = 1000000;
+
+    uint16_t aheadTimeUs = OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_US;
+    uint32_t busSpeedHz  = otPlatRadioGetBusSpeed(&aInstance);
+
+    if (busSpeedHz > 0)
+    {
+        aheadTimeUs +=
+            static_cast<uint16_t>((kWakeupFrameWeight * kBitsPerByte * kUsPerSecond + busSpeedHz - 1) / busSpeedHz);
+    }
+
+    return aheadTimeUs;
+}
+
+TimeMicro GetRadioNow(Instance &aInstance) { return TimeMicro(static_cast<uint32_t>(otPlatRadioGetNow(&aInstance))); }
+
+} // namespace
+
+RegisterLogModule("WakeupTxSched");
+
+WakeupTxScheduler::WakeupTxScheduler(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mTxTimeUs(0)
+    , mTxEndTimeUs(0)
+    , mTxRequestAheadTimeUs(CalcTxRequestAheadTimeUs(aInstance))
+    , mTimer(aInstance)
+    , mIsRunning(false)
+{
+}
+
+Error WakeupTxScheduler::WakeUp(const Mac::ExtAddress &aWedAddress, uint16_t aIntervalUs, uint16_t aDurationMs)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(!mIsRunning, error = kErrorInvalidState);
+
+    mWedAddress  = aWedAddress;
+    mTxTimeUs    = TimerMicro::GetNow() + mTxRequestAheadTimeUs;
+    mTxEndTimeUs = mTxTimeUs + aDurationMs * 1000 + aIntervalUs;
+    mIntervalUs  = aIntervalUs;
+    mIsRunning   = true;
+
+    LogInfo("Started wake-up sequence to %s", aWedAddress.ToString().AsCString());
+
+    ScheduleNext(/* aIsFirstFrame */ true);
+
+exit:
+    return error;
+}
+
+#if OPENTHREAD_CONFIG_RADIO_LINK_IEEE_802_15_4_ENABLE
+
+Mac::TxFrame *WakeupTxScheduler::PrepareWakeupFrame(Mac::TxFrames &aTxFrames)
+{
+    Mac::TxFrame      *frame = nullptr;
+    Mac::Address       target;
+    Mac::Address       source;
+    TimeMicro          radioTxUs;
+    uint32_t           rendezvousTimeUs;
+    Mac::ConnectionIe *connectionIe;
+
+    VerifyOrExit(mIsRunning);
+
+    target.SetExtended(mWedAddress);
+    source.SetExtended(Get<Mac::Mac>().GetExtAddress());
```

Comment:
Can consider removing `target` and `source` variables?
They are only uses below when `GenerateWakeupFrame()` is called. I think we can do this:

```c++
    VerifyOrExit(frame->GenerateWakeupFrame(Get<Mac::Mac>().GetPanId(),mWedAddress, Get<Mac::Mac>().GetExtAddress()) == kErrorNone, frame = nullptr);

Reply:
This doesn't seem possible as `Mac::Address` doesn't have non-default constructors that take `Mac::ExtAddress`.

Reply:
Yes. I see. You are right. Thanks.